File: balboa-service-jms/src/test/java/com/socrata/balboa/jms/BalboaJmsTest.java
package com.socrata.balboa.jms;

import org.junit.Test;

import java.util.List;

import static org.junit.Assert.*;

/**
 * Created by jonathan on 2/25/16.
 */
public class BalboaJmsTest {

    @Test
    public void testParseServers() throws Exception {
        String [] args = new String[3];
        args[0] = "5";
        args[1] = null;
        args[2] = "metrics2012";
        args[1] = "failover:(uri1,uri2)?transportOptions&nestedURIOptions";
        List<String> servers = BalboaJms.parseServers(args);
        assertEquals(1, servers.size());
        args[1] = "failover:(uri1,uri2)?transportOptions&nestedURIOptions,failover:(uri3,uri4)?transportOptions&nestedURIOptions";
        servers = BalboaJms.parseServers(args);
        assertEquals(2, servers.size());

        args = new String[4];
        args[0] = "5";
        args[1] = "failover:(uri1,uri2)?transportOptions&nestedURIOptions";
        args[2] = "failover:(uri3,uri4)?transportOptions&nestedURIOptions";
        args[3] = "metrics2012";
        servers = BalboaJms.parseServers(args);
        assertEquals(2, servers.size());
        assertEquals("failover:(uri1,uri2)?transportOptions&nestedURIOptions&soTimeout=15000&soWriteTimeout=15000", servers.get(0));
        assertEquals("failover:(uri3,uri4)?transportOptions&nestedURIOptions&soTimeout=15000&soWriteTimeout=15000", servers.get(1));


        args[1] = "failover:(uri1,uri2)";
        args[2] = "failover:(uri3,uri4)";
        servers = BalboaJms.parseServers(args);
        assertEquals(2, servers.size());
        assertEquals("failover:(uri1,uri2)?soTimeout=15000&soWriteTimeout=15000", servers.get(0));
        assertEquals("failover:(uri3,uri4)?soTimeout=15000&soWriteTimeout=15000", servers.get(1));

        args[1] = "failover:(uri1,uri2)";
        args[2] = "failover:(uri3,uri4)?soTimeout=15000&soWriteTimeout=15000";
        servers = BalboaJms.parseServers(args);
        assertEquals(2, servers.size());
        assertEquals("failover:(uri1,uri2)?soTimeout=15000&soWriteTimeout=15000", servers.get(0));
        assertEquals("failover:(uri3,uri4)?soTimeout=15000&soWriteTimeout=15000", servers.get(1));

    }
}

File: balboa-service-jms/src/test/scala/com/socrata/balboa/jms/BalboaJmsTest.scala
package com.socrata.balboa.jms

import com.typesafe.config.Config
import org.scalatest.{MustMatchers, WordSpec}
import org.scalatest.mock.MockitoSugar
import org.mockito.Mockito._
import com.socrata.balboa.jms.BalboaJmsTest.configFromTimeouts

class BalboaJmsTest extends WordSpec with MustMatchers {

  "BalboaJms::parseServers" should {
    "parse a single argument as one or more servers" in {
      {
      val args = Array(
        "failover:(uri1,uri2)?transportOptions&nestedURIOptions")
      val servers = BalboaJms.parseServers(args, configFromTimeouts())
      servers.size must be (1)
      }

      {
      val args = Array(
        "failover:(uri1,uri2)?transportOptions&nestedURIOptions," +
        "failover:(uri3,uri4)?transportOptions&nestedURIOptions")
      val servers = BalboaJms.parseServers(args, configFromTimeouts())
      servers.size must be (2)
      }
    }

    "produce servers from n arguments" in {
      {
      val args = Array(
        "failover:(uri1,uri2)?transportOptions&nestedURIOptions",
        "failover:(uri3,uri4)?transportOptions&nestedURIOptions")
      val servers = BalboaJms.parseServers(args, configFromTimeouts())
      servers.size must be (2)
      servers.get(0) must be ("failover:(uri1,uri2)?transportOptions&" +
        "nestedURIOptions&soTimeout=15000&soWriteTimeout=15000")
      servers.get(1) must be ("failover:(uri3,uri4)?transportOptions&" +
        "nestedURIOptions&soTimeout=15000&soWriteTimeout=15000")
      }

      {
      val args = Array(
        "failover:(uri1,uri2)",
        "failover:(uri3,uri4)")
      val servers = BalboaJms.parseServers(args, configFromTimeouts())
      servers.size must be (2)
      servers.get(0) must be (
        "failover:(uri1,uri2)?soTimeout=15000&soWriteTimeout=15000")
      servers.get(1) must be (
        "failover:(uri3,uri4)?soTimeout=15000&soWriteTimeout=15000")
      }

      {
      val args = Array(
        "failover:(uri1,uri2)",
        "failover:(uri3,uri4)?soTimeout=15000&soWriteTimeout=15000")
      val servers = BalboaJms.parseServers(args, configFromTimeouts())
      servers.size must be (2)
      servers.get(0) must be (
        "failover:(uri1,uri2)?soTimeout=15000&soWriteTimeout=15000")
      servers.get(1) must be (
        "failover:(uri3,uri4)?soTimeout=15000&soWriteTimeout=15000")
      }
    }

    "read timeouts from config" in {
      val args = Array(
        "failover:(uri1,uri2)?soWriteTimeout=15000",
        "failover:(uri3,uri4)?soTimeout=15000&soWriteTimeout=15000")
      val servers = BalboaJms.parseServers(args,
        configFromTimeouts(soTimeout = 30000))
      servers.size must be (2)
      servers.get(0) must be (
        "failover:(uri1,uri2)?soWriteTimeout=15000&soTimeout=30000")
      servers.get(1) must be (
        "failover:(uri3,uri4)?soTimeout=15000&soWriteTimeout=15000")
    }
  }
}

object BalboaJmsTest extends MockitoSugar {
  val soTimeoutPath = "activemq.sotimeout"
  val soWriteTimeoutPath = "activemq.sowritetimeout"

  def configFromTimeouts(soTimeout: Int = 15000, soWriteTimeout: Int = 15000): Config = {
    val conf = mock[Config]

    def hasPathAndConf(path: String, int: Int) = {
        when(conf.hasPath(path)).thenReturn(true)
        when(conf.getInt(path)).thenReturn(int)
    }

    hasPathAndConf(soTimeoutPath, soTimeout)
    hasPathAndConf(soWriteTimeoutPath, soWriteTimeout)

    conf
  }
}
